submariner-operator/issue-149-node-placement.md

```diff
@@ -0,0 +1,94 @@
+# Node Placement
+
+[Add support for nodeSelector and toleration scheduling for submariner pods](https://github.com/submariner-io/enhancements/issues/149)
+
+## Summary
+
+Submariner deploys many different agents as part of its core deployment. Except for Gateway, globalnet agent and
+metrics proxy, which are deployed on designated gateway nodes, there is no way for users to control or limit where rest
+of the agent pods are deployed. Some customers want to run these agents on specific nodes for billing or networking
+reasons.
+
+## Proposal
+
+This enhancement proposes adding a new field to `Submariner` and `ServiceDiscovery` for users to configure node
+scheduling. This configuration will then be applied to pods of following submariner agents:
+
+* submariner-operator
+* submariner-lighthouse-agent
+* submariner-lighthouse-coredns
+
+Following agents will be excluded because these are run on Gateway nodes. Users already control which nodes get labeled
+as gateways and ideally we don't want to run anything else on these nodes.
+
+* submariner-gateway
+* submariner-globalnet
+* submariner-metrics-proxy
+
+## Design Details
+
+Add a new `NodePlacement` field to `Submariner` and `ServiceDiscovery` specs.
+
+```Go
+// NodePlacement describes node scheduling configuration for the pods.
+type NodePlacement struct {
+    // NodeSelector defines which Nodes the Pods are scheduled on. The default is an empty list.
+    // +optional
+    NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+    // Tolerations is attached by pods to tolerate any taint that matches
+    // the triple <key,value,effect> using the matching operator <operator>.
+    // The default is an empty list.
+    // +optional
+    Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+}
+```
+
+```Go
+type SubmarinerSpec struct {
+	...
+	...
+	// NodePlacement enables explicit control over the scheduling of the submariner pods.
+	// +optional
+	NodePlacement NodePlacement `json:"nodePlacement,omitempty"`
+}
+```
+
+```Go
+type SubmarinerDiscoverySpec struct {
+	...
+	...
+	// NodePlacement enables explicit control over the scheduling of the submariner pods.
+	// +optional
+	NodePlacement NodePlacement `json:"nodePlacement,omitempty"`
+}
+```
+
+Add option to `subctl join` to specify node selectors and tolerations.
+
+```bash
+subctl join --node-selectors LABEL_KEY_1=LABEL_VALUE_1,...,LABEL_KEY_N=LABEL_VALUE_N
+subctl join --node-tolerations KEY_1=VAL_1:TAINT_EFFECT_1,...,KEY_N=VAL_N:TAINT_EFFECT_N
+```
+
+### Pros
+* Provides easy way for users to configure node scheduling
+
+### Cons
+* None
+
+### Backward Compatibility
+
+None
+
+### Alternatives
+
+None
```

Comment:
Mention that users can’t currently add their own tolerations etc. since the operator always rewrites the relevant K8s objects and any user changes are lost. (So taking user changes into account would be an alternative approach.)